### PR TITLE
[FSDP()] Fix `fully_shard` fwd hook registration

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90201 [FSDP()] Fix `fully_shard` fwd hook registration**

I need to rebase later after Shen's PRs land.

The idea is to only register the pre/post-forward hook on the _root modules_ among the modules that consume a `FlatParameter`. (Yes, the term _root module_ is heavily overloaded. We may want to clarify that at some point. Here, _root_ is being used in the graph sense, meaning parent-less, and the scope is only among the modules consuming a `FlatParameter`.)

This avoids unnecessary pre/post-forward hooks running, which would lead to errors because the unshard is not truly idempotent.

### CODE REVIEW

File: torch/distributed/fsdp/_init_utils.py [hunk 1]
```diff
@@ -175,10 +184,6 @@ def _init_runtime_state(
     state._pre_forward_handles = _pre_forward_handles
     _post_forward_handles: List[RemovableHandle] = []
     state._post_forward_handles = _post_forward_handles
-    _module_to_handles: Dict[
```
Comment (awgu) [inline torch/distributed/fsdp/_init_utils.py:178]: Remove redundant definition that slipped through

File: torch/distributed/fsdp/_common_utils.py [hunk 2]
```diff
@@ -200,3 +200,27 @@ def _assert_in_training_states(
             print(f"ERROR: {msg}")
             traceback.print_stack()
         raise ValueError(msg)
+
+
+def _get_root_modules(modules: Set[nn.Module]) -> Set[nn.Module]:
+    """
+    Returns:
+        Set[nn.Module: The subset of ``modules`` that are root modules (i.e.
```
Comment (mrshenli) [inline torch/distributed/fsdp/_common_utils.py:208]: missing "]"?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for promptly fixing this!

File: torch/distributed/fsdp/flat_param.py [hunk 3]
```diff
@@ -265,6 +268,7 @@ def _init_metadata(
         self._modules = set(pi.module for pi in self._param_infos).union(
             set(spi.module for spi in self._shared_param_infos)
         )
+        self._root_modules = _get_root_modules(self._modules)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:271]: nit: might be useful to add a comment describing why this property is needed (same goes for other properties as well).

**Review by rohan-varma — APPROVED** [review]
LGTM overall, but could we add a unittest that would've captured this issue?

**Comment by awgu** [comment]
> LGTM overall, but could we add a unittest that would've captured this issue?

Fortunately, Shen added the failing unit test in his PR (and the PR below this one patched that), so we have a unit test for this :)

File: torch/distributed/fsdp/flat_param.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:274]: Added explanation in the `FlatParameter` docstring

File: torch/distributed/fsdp/_common_utils.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/_common_utils.py]: Fixed

**Pull request closed without merging**